test/functional/saslprep_tests.js

```diff
@@ -0,0 +1,94 @@
+'use strict';
+
+const setupDatabase = require('./shared').setupDatabase;
+const withClient = require('./shared').withClient;
+
+describe('SASLPrep', function() {
+  // Step 4
+  // To test SASLprep behavior, create two users:
+  // username: "IX", password "IX"
+  // username: "u2168" (ROMAN NUMERAL NINE), password "u2163" (ROMAN NUMERAL FOUR)
+  // To create the users, use the exact bytes for username and password without SASLprep or other normalization and specify SCRAM-SHA-256 credentials:
+  // db.runCommand({createUser: 'IX', pwd: 'IX', roles: ['root'], mechanisms: ['SCRAM-SHA-256']}) db.runCommand({createUser: 'u2168', pwd: 'u2163', roles: ['root'], mechanisms: ['SCRAM-SHA-256']})
+  // For each user, verify that the driver can authenticate with the password in both SASLprep normalized and non-normalized forms:
+  // User "IX": use password forms "IX" and "Iu00ADX"
+  // User "u2168": use password forms "IV" and "Iu00ADV"
+  // As a URI, those have to be UTF-8 encoded and URL-escaped, e.g.:
+  // mongodb://IX:IX@mongodb.example.com/admin
+  // mongodb://IX:I%C2%ADX@mongodb.example.com/admin
+  // mongodb://%E2%85%A8:IV@mongodb.example.com/admin
+  // mongodb://%E2%85%A8:I%C2%ADV@mongodb.example.com/admin
+
+  const users = [
+    {
+      username: 'IX',
+      password: 'IX',
+      mechanisms: ['SCRAM-SHA-256']
+    },
+    {
+      username: '\u2168',
+      password: '\u2163',
+      mechanisms: ['SCRAM-SHA-256']
+    }
+  ];
+
+  before(function() {
+    return setupDatabase(this.configuration);
+  });
+
+  before(function() {
+    return withClient(this.configuration.newClient(), client => {
```

Comment:
hmm I wonder if this `withClient` is actually worth it? I count this file as creating eight clients. You should test test time length with and without `withClient`, the overhead for setting up a TCP connection, handshaking, etc is non-trivial. I'm all about coming up with a pattern that effectively sets up and tears down clients, but I'm not convinced this is the way yet

Reply:
I believe that, since we don't actually tear down clients properly such that they can be reused, we need a separate client to test different auth credentials each time.

Reply:
good point, objection retracted